Mathlib/Order/SuccPred/Tree.lean

```diff
@@ -0,0 +1,230 @@
+
+end IsPredArchimedean
+
+/--
+The type of rooted trees.
+-/
+structure RootedTree where
+  /-- The type representing the elements in the tree. -/
+  α : Type*
+  /-- The type should be a `SemilatticeInf`,
+    where `inf` is the least common ancestor in the tree. -/
+  [semilatticeInf : SemilatticeInf α]
+  /-- The type should have a bottom, the root. -/
+  [orderBot : OrderBot α]
+  /-- The type should have a predecessor for every element, its parent. -/
+  [predOrder : PredOrder α]
+  /-- The predecessor relationship should be archimedean. -/
+  [isPredArchimedean : IsPredArchimedean α]
+
+attribute [coe] RootedTree.α
+
+instance : CoeSort RootedTree Type* := ⟨RootedTree.α⟩
+
+instance (t : RootedTree) : SemilatticeInf t := t.semilatticeInf
+instance (t : RootedTree) : PredOrder t := t.predOrder
+instance (t : RootedTree) : OrderBot t := t.orderBot
+instance (t : RootedTree) : IsPredArchimedean t := t.isPredArchimedean
+
+/--
+A subtree is represented by its root, therefore this is a type synonym.
+-/
+def SubRootedTree (t : RootedTree) : Type* := t
+
+/--
+The root of a `SubRootedTree`.
+-/
+def SubRootedTree.root {t : RootedTree} (v : SubRootedTree t) : t := v
+
+/--
+The `SubRootedTree` rooted at a given node.
+-/
+def RootedTree.subtree (t : RootedTree) (r : t) : SubRootedTree t := r
+
+@[simp]
+lemma RootedTree.root_subtree (t : RootedTree) (r : t) : (t.subtree r).root = r := rfl
+
+@[simp]
+lemma RootedTree.subtree_root (t : RootedTree) (v : SubRootedTree t) : t.subtree v.root = v := rfl
+
+@[ext]
+lemma SubRootedTree.ext {t : RootedTree} {v₁ v₂ : SubRootedTree t}
+    (h : v₁.root = v₂.root) : v₁ = v₂ := h
+
+instance (t : RootedTree) : SetLike (SubRootedTree t) t where
+  coe v := Set.Ici v.root
+  coe_injective' a₁ a₂ h := by
+    simpa only [Set.Ici_inj, ← SubRootedTree.ext_iff] using h
+
+lemma SubRootedTree.mem_iff {t : RootedTree} {r : SubRootedTree t} {v : t} :
+    v ∈ r ↔ r.root ≤ v := Iff.rfl
+
+/--
+The coercion from a `SubRootedTree` to a `RootedTree`.
+-/
+@[coe, reducible]
+noncomputable def SubRootedTree.coeTree {t : RootedTree} (r : SubRootedTree t) : RootedTree where
+  α := Set.Ici r.root
+
+noncomputable instance (t : RootedTree) : CoeOut (SubRootedTree t) RootedTree :=
+  ⟨SubRootedTree.coeTree⟩
+
+@[simp]
+lemma SubRootedTree.bot_mem_iff {t : RootedTree} (r : SubRootedTree t) :
+    ⊥ ∈ r ↔ r.root = ⊥ := by
+  simp [mem_iff]
+
+/--
+All of the immediate subtrees of a given rooted tree.
+-/
+def RootedTree.subtrees (t : RootedTree) : Set (SubRootedTree t) :=
+  {x | IsAtom x.root}
+
+variable {t : RootedTree}
+
+lemma SubRootedTree.root_ne_bot_of_mem_subtrees (r : SubRootedTree t) (hr : r ∈ t.subtrees) :
+    r.root ≠ ⊥ := by
+  simp only [RootedTree.subtrees, Set.mem_setOf_eq] at hr
+  exact hr.1
+
+lemma RootedTree.subtrees_inf_eq_bot_iff {t₁ t₂ : SubRootedTree t}
+    (ht₁ : t₁ ∈ t.subtrees) (ht₂ : t₂ ∈ t.subtrees) (v₁ v₂ : t) (h₁ : v₁ ∈ t₁)
+    (h₂ : v₂ ∈ t₂) :
+    v₁ ⊓ v₂ = ⊥ ↔ t₁ ≠ t₂ where
```

Comment:
```suggestion
    Disjoint v₁ v₂ ↔ t₁ ≠ t₂ where
```
this should work?

Reply:
Updated, although I'm not sure about the new name `mem_subtrees_disjoint_iff`

Reply:
`disjoint_iff_of_mem_subtrees`, maybe? or `disjoint_iff_ne_of_mem_subtrees`